pkg/scheduler/framework/framework.go

```diff
@@ -143,8 +174,738 @@ func (f *framework) EventRecorder() record.EventRecorder {
+		// Note that the op would fail if the policy snapshot is not the latest, so that consistency is
+		// preserved.
+		if err := f.updatePolicySnapshotStatus(ctx, policy, refreshedSchedulingDecisions, newSchedulingCondition); err != nil {
+			klog.ErrorS(err, "failed to update policy snapshot status during downscaling", "schedulingPolicySnapshot", schedulingPolicySnapshotRef)
+			return ctrl.Result{}, err
+		}
+
+		// Return immediately as there are no more bindings for the scheduler to scheduler at this moment.
+		return ctrl.Result{}, nil
+	}
+
+	// Check if the scheduler needs to take action; a scheduling cycle is only needed if
+	// * the policy is of the PickAll type; or
+	// * the policy is of the PickN type, and currently there are not enough number of bindings.
+	if !shouldSchedule(policy, numOfClusters, len(active)+len(creating)) {
+		// No action is needed; however, a status refresh might be warranted.
+		//
+		// This is needed as a number of situations (e.g., POST/PUT failures) may lead to inconsistencies between
+		// the decisions added to the policy snapshot status and the actual list of bindings.
+		klog.V(2).InfoS("No scheduling is needed", "schedulingPolicySnapshot", schedulingPolicySnapshotRef)
+
+		// Refresh scheduling decisions.
+		refreshedSchedulingDecisions := refreshSchedulingDecisionsFrom(policy, active, creating)
+		// Prepare new scheduling condition.
+		//
+		// In this case, since no action is needed, the scheduler considers the policy to be fully scheduled.
+		newSchedulingCondition := fullyScheduledCondition(policy)
+
+		// Update the policy snapshot status.
+		//
+		// Note that the op would fail if the policy snapshot is not the latest, so that consistency is
+		// preserved.
+		if err := f.updatePolicySnapshotStatus(ctx, policy, refreshedSchedulingDecisions, newSchedulingCondition); err != nil {
+			klog.ErrorS(err, "failed to update policy snapshot status (no scheduling needed)", "schedulingPolicySnapshot", schedulingPolicySnapshotRef)
+			return ctrl.Result{}, err
+		}
+
+		// Return immediate as there no more bindings for the scheduler to schedule at this moment.
+		return ctrl.Result{}, nil
+	}
+
+	// The scheduler needs to take action; enter the actual scheduling stages.
+	klog.V(2).InfoS("Scheduling is needed; entering scheduling stages", "schedulingPolicySnapshot", schedulingPolicySnapshotRef)
+
+	// Prepare the cycle state for this run.
+	//
+	// Note that this state is shared between all plugins and the scheduler framework itself (though some fields are reserved by
+	// the framework). These resevered fields are never accessed concurrently, as each scheduling run has its own cycle and a run
+	// is always executed in one single goroutine; plugin access to the state is guarded by sync.Map.
+	state := NewCycleState()
+
+	// Calculate the batch size.
+	//
+	// Note that obsolete bindings are not counted.
+	state.desiredBatchSize = int(*policy.Spec.Policy.NumberOfClusters) - len(active) - len(creating)
```

Comment:
NumberOfClusters is optional, what if it is nil?

Reply:
I have split the flow for PickAll and PickN for better clarity 

For this specific point, it is a leftover from earlier code where it is not yet certain whether numOfClusters will be annotated; it should use the extracted numOfClusters instead.